### ISSUE TITLE
FIP divergence on Zen

### ISSUE THREAD

**Issue by rocallahan** [mozilla/rr#2698]
Trying to replay a Zen-recorded trace on an Intel CPU we observed a case where AMD's FIP field of an `fnstenv` was zero while the field was nonzero on Intel.

**Comment by rocallahan** [comment]
I found this: https://patchwork.kernel.org/patch/8390831/

> The hardware may not write the FIP/FDP fields with a XSAVE*
instruction.  e.g., with XSAVEOPT/XSAVES if the state hasn't changed
or on AMD CPUs when a floating point exception is not pending.

**Comment by rocallahan** [comment]
https://github.com/mozilla/rr/issues/2687#issuecomment-699451979 confirms this behavior.

**Comment by rocallahan** [comment]
We should probably extend the `x87env` test to nail down the exact behaviour.

**Comment by rocallahan** [comment]
According to https://github.com/mozilla/rr/issues/2687#issuecomment-699463432 `x87env` produces 0 for the FSTENV FIP when run under rr, and the right result when run normally. This is surprising since the instructions sequences executed are the same and there's no ptrace interrupt between the x87 instruction and FSTENV when run under rr.

A possible theory is that when run normally the process is in an x87 exception state when it reaches our inline assembly fragment, and this causes the FIP address to be stored, but under rr for some reason the x87 exception state is cleared earlier. Then the statement that AMD hardware does not write FIP when there is no exception state would mean we see a 0 FIP value.

**Comment by rocallahan** [comment]
That theory might also explain why `x87env` run outside of rr, with https://github.com/mozilla/rr/commit/551b00297def38a32a140a6c54517263b960dd61, returns 0 for the XSAVE64 --- because FSTENV may be clearing the exception state.

> The FSTENV instruction checks for and handles any pending unmasked floating-point exceptions before storing the FPU environment; the FNSTENV instruction does not.

**Comment by rocallahan** [comment]
I've been analyzing a version of x87env under gdb on @glandium's machine.

If I set a breakpoint on the call to sched_yield, stop there in gdb, then continue, I get FIP=0.
If I configure the breakpoint to continue immediately (using `command ...`, `continue`, `end`), then I get the FIP of the `fld1` as expected.
If I configure the breakpoint to print registers using `info reg` or `info reg float` before continuing, I get the FIP of `fld1`.
If I configure the breakpoint to print registers using `info reg save` (which I guess uses XSAVE under the hood) I get FIP=0!

**Comment by rocallahan** [comment]
> there's no ptrace interrupt between the x87 instruction and FSTENV when run under rr.

This is incorrect. There is a PATCH_SYSCALL event between those instructions.

**Comment by rocallahan** [comment]
This appears to be a timing issue.

If I configure the breakpoint to `!sleep 1`, I reliably get FIP=0.
If I configure the breakpoint to `!sleep 0`, I sometimes get FIP=0 and otherwise `fld1`.

:-(

**Comment by rocallahan** [comment]
We probably should worry about FDP as well.

**Comment by rocallahan** [comment]
One thing we could try is to clear FIP/FDP at every rr trap on Zen. This would make x87env fail on Zen but at least it would replay correctly I guess.

**Comment by rocallahan** [comment]
... which is good enough since no-one really cares about the values of FIP/FDP.

**Comment by rocallahan** [comment]
We would want to record in the trace a flag that we have done this, so that we reliably do it on replay no matter which CPU we replay on.

**Comment by rocallahan** [comment]
One possible explanation for this timing behavior is that the kernel scheduler at some point decides to schedule a different userspace process on that core, which requires XSAVE/XRSTOR, and the XSAVE doesn't save FIP because there is no x87 exception pending, so when we switch back to our process the FIP is set to zero.

**Comment by rocallahan** [comment]
It seems like a bug that XSAVE would be leaky in this way.

**Comment by rocallahan** [comment]
```
#include <assert.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

extern char _x87_instruction;

int main(int argc, char** argv) {
  uint32_t fstenv_buf[256];
  struct timespec ts = { 0, 0 };

  if (argc == 1) {
    printf("Usage: %s <nanoseconds>", argv[0]);
    return 0;
  }
  ts.tv_nsec = atoi(argv[1]);

  __asm__ volatile("_x87_instruction: fld1\n");
  puts("Sleeping...");
  nanosleep(&ts, NULL);
  __asm__ volatile("fstenv %0\n"
                   : "=m"(fstenv_buf)
                   :
                   : "memory");
  printf("FIP=0x%x\n", fstenv_buf[3]);
  assert(fstenv_buf[3] == (uint32_t)(uintptr_t)&_x87_instruction);
  puts("EXIT-SUCCESS");
  return 0;
}
```
Compiling this as `test`, on @glandium's machine, running `test 50000000` (i.e. 50ms sleep) fails pretty often.

**Comment by rocallahan** [comment]
I can reproduce similar behaviour with a busyloop. This suggests there is nondeterminism potentially visible where kernel scheduling decisions can clear FIP at any time :-(.

**Comment by rocallahan** [comment]
If these hypotheses are correct, I think probably the best workaround we can do is to clear FIP/FDP on every rr trap on Zen (perhaps only if no x87 exception is pending in the saved state). We'd still be vulnerable to application code doing an x87 instruction followed by an `fstenv`/`xsave` with the kernel clearing FIP in between but no rr trap in between --- this wouldn't even replay reliably on the same machine. But given x87 usage is rare, that scenario should be very rare.

**Comment by rocallahan** [comment]
One interesting quirk, but I think unrelated to the main issue here, is that if you set a breakpoint at the `fld1`instruction itself with gdb, stop there, then `continue`, the FIP value reported is somewhere near the beginning of `_start` (nothing x87 related). gdb handles the `continue` by doing a `PTRACE_SINGLESTEP` to get past the breakpoint followed by `PTRACE_CONT`, so I guess it's something to do with the way the kernel handles singlestepping, or maybe it's a weird processor quirk. This seems to happen the same way on Zen and Intel though so I don't think it will affect rr.

**Comment by rocallahan** [comment]
If we clear FIP/FDP on every rr event when recording on Zen, that will make divergence much less likely during replay on Zen and Intel.

The other interesting scenario is recording on Intel and replaying on Zen. We could clear FIP/FDP on Intel too, at the expense of two extra syscalls per rr event that aren't really needed if you aren't going to replay on Zen. I'm not sure what we should do about that. Anyway it's lower priority.

**Comment by rocallahan** [comment]
BTW for porting traces between Zen and Intel CPUID faulting comes into play. No CPUID faulting on Zen so the Intel record or replay machine must have CPUID faulting.

**Comment by rocallahan** [comment]
Some more detailed analysis.

### Record on Zen, replay on Zen

The problem happens when 1) the application executes an x87 instruction, 2) the kernel clears FIP (often during a trap to rr), 3) an application XSAVE with x87 exception pending or a FSTENV leaks the FIP value into memory, and 4) that value leaks into registers visible at an rr event (or the application uses it, but let's ignore that as implausible). This can happen during record or replay.

Clearing FIP on every rr event during recording and replay reduces the window for step 3 from "anywhere on that thread after step 2" to "before recording the next rr event".

### Record on Zen, replay on Intel

Same problem as above, but also, instead of steps 2 and 3 the application could just do an XSAVE with no x87 exception pending, which stores FIP 0 on Zen but not on Intel. 

Clearing FIP on every rr event during recording and replay should significantly reduce the number of possible XSAVEs that would produce different results, because there has to be an x87 instruction executed in the rr event leading up to the XSAVE. This should help significantly because I expect step 4 to be rare --- a few diverging XSAVES are unlikely to cause problems.

### Record on Intel, replay on Zen

If we clear FIP during recording on Intel, this is very much like the previous case. If we don't, any XSAVE or FSTENV that leaks FIP into rr-observable state is likely to cause replay failure since FIP will probably be zero on Zen but not on Intel if even one x87 instruction has been executed.

### Binary-instrumentation-enabled replay

With binary instrumentation available during replay (e.g. in Pernosco) the situation changes a bit for the cross-vendor cases. We can emulate the Intel behavior of XSAVE on Zen, and the Zen behavior on Intel. It doesn't help with kernel asynchronously clearing FIP though, so we'd be back to the same reliability as "record and replay on the same Zen machine".

**Comment by rocallahan** [comment]
BTW this is relevant to FDP: https://robert.ocallahan.org/2018/05/rr-trace-portability-x87-data-pointer.html

**Comment by rocallahan** [comment]
I've checked the CPU behavior on my Skylake laptop and @glandium 's machine:

On Skylake (and, I assume, any later processor with FDP_EXCPTN_ONLY), FDP is only set when an x87 instruction with a memory operand executes and raises an unmasked exception. Zen sets FDP whenever an x87 instruction with a memory operand executes. (I assume this matches pre-Skylake Intel.) Execution of other x87 instructions does not seem to clear FDP.

I guess we should add a new trace flag to capture this behavior in case we want to reproduce it with instrumentation. Might be useful for diagnostics too.

The Zen XSAVE behavior is that FIP/FDP is only written to memory if there is a pending x87 exception. The actual FIP/FDP values are still there, you can read them later with an FNSTENV.

FIP seems to be set on every x87 instruction no matter what.
